### ISSUE TITLE
[feature request] Set limit on GPU memory use

### ISSUE THREAD

**Issue by sandias42** [pytorch/pytorch#18626]
## 🚀 Feature

Allow user to easily specify a fraction of the GPU memory to use.

## Motivation

I recently switched from tensorflow to pytorch for what I saw as greater flexibility and user control. However, I have been recently frustrated by the inability to specify a cap on the fraction of GPU memory my pytorch process should be using. I have what I think is a fairly standard use case in performing hyper-parameter search by running multiple independent training processes in parallel (there is a whole ecosystem of packages for this). A modern GPU is large enough to train 4-8+ models of my size, but very rarely a configuration is selected which uses almost my full GPU memory. Instead of that memory-hungry process failing with an OOM by using more than its share (the sensible behavior), instead all the other training processes die, and as my script automatically generates more to take their place, they fail too, leading to a catastrophic global failure. 

There are additional use cases like shared servers, training models with small batch sizes for statistical efficiency, and using very large GPUs effectively (like modern Voltas). 

In tensorflow, this can be done simply by passing a `ConfigProto` with `gpu_memory_fraction=x`. 

There seems to be a pain point for others as well (see Additional context). I know that for me, finding a way to do this will probably make the difference on switching back to TF. 

## Pitch

It would be awesome if an option like gpu_memory_fraction could be set somewhere in the pytorch flow. My current recommendation would be to allow an optional arg or kwarg to torch.device, like the index argument, which would specify the fraction of GPU memory to use. If more than `gpu_mem_fraction * total_gpu_mem` is attempted to be allocated, raise OOM. 

## Alternatives

It seems like you might be able to do some kind of hacky work around using https://pytorch.org/docs/stable/cuda.html#torch.cuda.max_memory_allocated and monitoring of each process being trained.

## Additional context

https://discuss.pytorch.org/t/how-to-set-a-limit-to-gpu-usage/7271
https://stackoverflow.com/questions/49529372/force-gpu-memory-limit-in-pytorch
https://discuss.pytorch.org/t/limiting-gpu-usage/7662

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

**Comment by aluo-x** [comment]
Wanted to support this feature request as well. 

For context, this could be used with the work done by Alibaba for kubernetes (See [Kubernetes #52757](https://github.com/kubernetes/kubernetes/issues/52757#issuecomment-464645334)). While there is probably a significant speed/latency penalty, it may still be beneficial for certain tasks.

**Comment by grnikh** [comment]
I have been looking for this feature for a while- one of my standard training workflows has been extremely unreliable in PyTorch (also recently switched from TF) due to gpu memory overflows leading to cascading failures. It would be very useful to me to neatly control the maximum gpu fraction even with a substantial performance reduction

**Comment by small-yellow-duck** [comment]
I would also be very useful to me to be able to control the portion of the gpu allocated to a process.

**Comment by alibabadoufu** [comment]
I am supporting as well.

**Comment by kealennieh** [comment]
This feature will be very useful.

**Comment by psinger** [comment]
Supporting this.

**Comment by lhty24** [comment]
Supporting this. This would be a very useful feature. I was actually surprised to find out that Pytorch doesn't have this.

**Comment by gloryVine** [comment]
I need this too.

**Comment by r734** [comment]
This would go with the [K8s Shared GPU Scheduler](https://github.com/AliyunContainerService/gpushare-scheduler-extender) like peanut butter with chocolate.

**Comment by feiwang3311** [comment]
same here. This feature would be useful

**Comment by LovelyQuantum** [comment]
Support! This feature will be especially useful when in production environment.

**Comment by Aaronzinhoo** [comment]
Support! This will allow for flexibility in workflows with multiple networks being used at once among all the other benefits mentioned above.

**Comment by gauthamkrishna-g** [comment]
Supporting this as well!

**Comment by birdx0810** [comment]
Support ++

**Comment by MarisaKirisame** [comment]
Is anyone working on this? I dont think it is very hard, one can add a line in CUDACachingAllocator's malloc, saying after the allocation is done, if it exceed the memory, immediately free it and error, or something like that.

**Comment by hermanius** [comment]
support++

**Comment by LieutenantChips** [comment]
++

**Comment by mmalghamdi** [comment]
support this feature.

**Comment by SpicySyntax** [comment]
+1 this could also be useful for model serving. E.g. sharing GPU for inference across multiple processes.

**Comment by Califax** [comment]
This would be really useful for sharing gpus especially in kubernetes clusters.

**Comment by dlrac** [comment]
Extremely useful feature...

**Comment by EvgenyM** [comment]
+1

**Comment by hzhwcmhf** [comment]
Support!  Extremely useful if I want to run several networks on a single GPU.

**Comment by JulianMar11** [comment]
Support +1

**Comment by adityarajagopal** [comment]
Would be great to have.

**Comment by lk123400** [comment]
Suppot +1

**Comment by lk123400** [comment]
Support + 1

**Comment by Mikhail-M** [comment]
Support +1

**Comment by wujushan** [comment]
++

**Comment by adrianosantospb** [comment]
Support +1

**Comment by suraj081** [comment]
support and want very badly..

**Comment by lamhoangtung** [comment]
I'm supporting this as well

**Comment by lihaoxiang1989** [comment]
I support this as well. We got multiple models to run which reserved less 10G GPU memory, but it was going to 20G when split into 5 processes

**Comment by neo133** [comment]
I'm supporting this feature

**Comment by n2cholas** [comment]
+1, this would be very useful!

**Comment by kepler62f** [comment]
Support +1

**Comment by ShangxuanWu** [comment]
Support +1

**Comment by mcvarer** [comment]
Support +1

**Comment by adodd202** [comment]
Support +1

**Comment by Hubes92** [comment]
support +1

**Comment by flocreate** [comment]
+1

**Comment by carlislewang** [comment]
+1

**Comment by archiducarmel** [comment]
Support this feature : +1

**Comment by ririya** [comment]
+1. Pytorch is allocating 2 times more memory than i need.

**Comment by brannondorsey** [comment]
It seems like there is overwhelming community interest in this feature, but there doesn't appear to be much response from the PyTorch development team. Could someone close to the PyTorch development process/team shed some light on whether this is being actively considered on an upcoming roadmap?

If not, is there a reason this change would be considered out of scope or de-prioritized?

**Comment by semaraugusto** [comment]
+1

**Comment by shepherd1530** [comment]
+1

**Comment by zou3519** [comment]
Tentatively bumping to high-pri based on user activity.

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/44742 will be setting a precedent to allow configurable knobs on the memory allocator; a GPU memory limit would easily fit into this framework

**Comment by mruberry** [comment]
cc @VitalyFedyunin

**Comment by ngimel** [comment]
there's a [`torch.cuda.set_per_process_memory_fraction`](https://github.com/pytorch/pytorch/blob/fe645fdfc726744315bbef1b7a01bc8a420a919c/torch/cuda/memory.py#L75) function that does that

**Comment by IgorStetsenko** [comment]
I got the message "AttributeError: module 'torch._C' has no attribute '_cuda_setMemoryFraction'". 

Help me, please.

**Comment by ngimel** [comment]
What pytorch version are you using?

**Comment by Shobhit20** [comment]
@ngimel I am getting the same error AttributeError: module 'torch.cuda' has no attribute 'set_per_process_memory_fraction' while using torch 1.4.0 
Can you tell the version beyond which this support is added? 
Also in our use case we are using detection model deployed using torchserve. Will that inherit this property if it is defined right at the start before loading the model

**Comment by ngimel** [comment]
It was recently added in #48172, so it's available only in nightly builds and will be available in 1.8 release.

**Comment by dreamflasher** [comment]
Pytorch 1.8 is released: https://pytorch.org/blog/pytorch-1.8-released/ – so this issue can be closed. It's finally there! :)
